File: labs/lab5/Dictionary.c
//Isai Lopez
//ilopezro
//Dustin Adams
//CMPS12B - Lab 5
//implementing a Dictionary ADT in C

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

//private types--------------------------

// NodeObj
typedef struct NodeObj{
   char* mKey;
   char* mValue; 
   struct NodeObj* mNext;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* key, char* value) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->mKey = key;
   N->mValue = value;
   N->mNext = NULL; 
   return(N);
}


// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// Dictionary Obj
typedef struct DictionaryObj{
	Node head;
   int numItems;
} DictionaryObj;

//find key 
Node findKey(Dictionary D, char* key){
	// Node temp = (*D)->head; 
	Node temp = D->head;  
	while(temp!=NULL)
	{
		if(strcmp(key, temp->mKey) ==0)
			return temp; 
		temp = temp->mNext; 
	}
	freeNode(&temp); 
	return NULL; 
} 



//public types ---------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj)); 
	assert(D!=NULL); 
	D->head = NULL; 
	D->numItems = 0; 
	return D; 
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
	if( pD!= NULL && *pD!=NULL){
		if( !isEmpty(*pD) ) makeEmpty(*pD); 
			free(*pD); 
		*pD = NULL; 
	}
}

// isEmpty()
// returns 1(true) if D is empty, 0(false) otherwise 
// pre: none 
int isEmpty(Dictionary D){
	return(D->numItems == 0); 
}

// size() 
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
	return(D->numItems); 
}

// lookup()
// returns the value v such that (k,v) is in D, or returns NULL if no such value v exists 
// pre: none
char* lookup(Dictionary D, char*k){
	if(findKey(D, k) != NULL)
	{
		Node temp = D->head; 
		while(temp != findKey(D, k))
		{
			temp = temp->mNext; 
		}
		return temp->mValue; 
	}
	else
		return NULL; 
}

//insert()
//inserts new (key,value) pair into D
// pre: lookup(D,k) == null
void insert(Dictionary D, char* k, char*v)
{
	if(lookup(D, k) != NULL)
	{
		fprintf(stderr, "Error: Cannot insert duplicate keys\n"); 
		exit(EXIT_FAILURE); 
	}
	// Node toAdd = newNode(k,v); 
	if(D->numItems == 0)
	{
		D->head = newNode(k,v); 
	}
	else
	{
		Node temp = D->head; 
		while(temp->mNext!= NULL)
		{
			temp = temp->mNext; 
		}
		temp->mNext = newNode(k,v);
		// freeNode(&temp);  
	}
	D->numItems++; 
}

//delete()
//deletes pair with the key K 
// pre: lookup(D,k) != null
void delete(Dictionary D, char* k)
{
	if(lookup(D, k) == NULL)
	{
		fprintf(stderr, "Error: cannot delete non-existent key\n");
		exit(EXIT_FAILURE); 
	}
	else
	{
		if(D->head == findKey(D, k))
		{
			Node temp = D->head; 
			D->head = temp->mNext;
			freeNode(&temp); 
		}
		else
		{
			Node temp = D->head; 
			while(temp->mNext != findKey(D, k))
			{
				temp = temp->mNext; 
			}
			Node toDelete = temp->mNext; 
			temp->mNext = temp->mNext->mNext; 
			freeNode(&toDelete); 
		}
		D->numItems--; 	
	}
}


//makeEmpty()
// re-sets D to the empty state
// pre: none
void makeEmpty(Dictionary D)
{
	while(D->numItems>0)
	{
		Node temp = D->head; 
		D->head = temp->mNext; 
		freeNode(&temp); 
		D->numItems--; 
	} 
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed out by out
void printDictionary(FILE* out, Dictionary D)
{
	Node N; 
	for(N=D->head; N!=NULL; N=N->mNext)
	{
		fprintf(out, "%s %s \n", N->mKey, N->mValue); 
	}
	freeNode(&N); 
}

File: labs/lab7/Dictionary.java
/**
 * Isai Lopez
 * ilopezro
 * Lab 7
 * Mar 12, 2018
 * (Doc Description)
 */

public class Dictionary implements DictionaryInterface {

	//Private Methods 

	// inner Node class w
	private class Node 
	{
		String mKey;
		String mValue; 
		Node mLeft;
		Node mRight; 

		private Node(String key, String value){
			mKey = key;
			mValue = value; 
			mLeft = null;
			mRight = null; 
		}
	}

	private Node mRoot;     // reference to root node
	private int mNumPairs;  // number of items in this Dictionary

	// findKey()
	// returns the Node containing key k in the subtree rooted at mRoot, or returns
	// NULL if no such Node exists
	private Node findKey(Node root, String key)
	{ 
		if(root == null ||key.compareTo(root.mKey) == 0) {
			return root; 
		}
		if(key.compareTo(root.mKey)< 0) {
			return findKey(root.mLeft, key); 
		}else {
			return findKey(root.mRight, key); 
		}
	}
	
	//findParent()
	//return the parent of N in the subtree rooted at R, or return null 
	//if N is equal to R 
	//pre: mRoot != null
	private Node findParent(Node N, Node R){
		Node parentNode = null; 
		if(N != R) {
			parentNode = R; 
			while(parentNode.mLeft!= N && parentNode.mRight!=N) {
				if(N.mKey.compareTo(parentNode.mKey) < 0)
					parentNode = parentNode.mLeft; 
				else
					parentNode = parentNode.mRight; 
			}
		}
		return parentNode; 
	}
	
	// findLeftmost()
	// returns the leftmost Node in the subtree rooted at mRoot, or NULL if mRoot is NULL.
	private Node findLeftmost(){
		Node leftMost = this.mRoot; 
		if(leftMost!=null) { for(; leftMost.mLeft !=null; leftMost = leftMost.mLeft); }
		return leftMost; 
	}
	
	// printInOrder()
	// prints the (key, value) pairs belonging to the subtree rooted at R in order
	// of increasing keys to file pointed to by out.
	private String printInOrder(Node root){
		String toString = ""; 
		if(root != null) {
			toString += printInOrder(root.mLeft); 
			toString += root.mKey + " " + root.mValue + "\n";
			toString += printInOrder(root.mRight); 
		}
		return toString; 
	}
	
	// deleteAll()
	// deletes all Nodes in the subtree rooted at N.
	private void deleteAll(){
		mRoot = null; 
		mNumPairs = 0; 
	}

	//Public Methods used from DictionaryInterface
	
	// constructor for Dictionary class
	public Dictionary()
	{
		mRoot = null;
		mNumPairs = 0;
	}
	
	// isEmpty()
	// returns 1 (true) if D is empty, 0 (false) otherwise
	// pre: none
	public boolean isEmpty() {
		return (mNumPairs==0);
	}

	// size()
	// returns the number of (key, value) pairs in D
	// pre: none
	public int size() {
		return mNumPairs;
	}

	// lookup()
	// returns the value v such that (k, v) is in D, or returns NULL if no 
	// such value v exists.
	// pre: none
	public String lookup(String key) {
		Node toFind = findKey(this.mRoot, key);
		if(toFind == null)
			return null; 
		else 
			return toFind.mValue; 
	}

	// insert()
	// inserts new (key,value) pair into D
	// pre: lookup(key)==NULL
	public void insert(String key, String value) throws DuplicateKeyException {
		Node toAdd, toTraverse, helper; 
		if(lookup(key) != null) {
			throw new DuplicateKeyException("Dictionary Error: cannot insert() duplicate key: " + key); 
		}
		toAdd = new Node(key, value); 
		helper = null; 
		toTraverse = this.mRoot; 
		while(toTraverse != null) {
			helper = toTraverse; 
			if(key.compareTo(toTraverse.mKey) < 0)
				toTraverse = toTraverse.mLeft; 
			else
				toTraverse = toTraverse.mRight; 
		}
		if(helper == null)
			this.mRoot = toAdd; 
		else if(key.compareTo(helper.mKey) < 0)
			helper.mLeft = toAdd; 
		else
			helper.mRight = toAdd; 
		this.mNumPairs++;
	}

	// delete()
	// deletes pair with the key k
	// pre: lookup(D, k)!=NULL
	public void delete(String key) throws KeyNotFoundException {
		Node toDelete, parentNode, fourthCase; 
		if(lookup(key) == null) {
			throw new KeyNotFoundException("Dictionary Error: cannot delete() non-existent key: " + key); 
		}
		toDelete = findKey(mRoot, key);
		//#1 case: no children 
		if(toDelete.mLeft == null && toDelete.mRight == null) {
			if(toDelete == mRoot) {
				mRoot = null; 
			}else {
				parentNode = findParent(toDelete, mRoot);
				if(parentNode.mRight ==toDelete) {
					parentNode.mRight = null; 
				}else {
					parentNode.mLeft = null; 
				}
			}
		}
		//#2 case: left child, but no right child
		else if(toDelete.mRight == null) {
			if(toDelete == mRoot) {
				mRoot = toDelete.mLeft; 
			}else {
				parentNode = findParent(toDelete, mRoot); 
				if(parentNode.mRight == toDelete) {
					parentNode.mRight = toDelete.mLeft; 
				}else {
					parentNode.mLeft = toDelete.mLeft; 
				}
			}
		}
		//#3 case: right child, but no left child
		else if(toDelete.mLeft == null) {
			if(toDelete == mRoot) {
				mRoot = toDelete.mRight; 
			}else {
				parentNode = findParent(toDelete, mRoot); 
				if(parentNode.mRight == toDelete) {
					parentNode.mRight = toDelete.mRight; 
				}else {
					parentNode.mLeft = toDelete.mRight; 
				}
			}
		}
		//case #4: two children if(toDelete.mLeft != null && toDelete.mRight != null)
		else {
			fourthCase = findLeftmost(); 
			toDelete.mKey = fourthCase.mKey; 
			toDelete.mValue = fourthCase.mValue; 
			parentNode = findParent(fourthCase, toDelete); 
			if(parentNode.mRight == fourthCase) {
				parentNode.mRight = fourthCase.mRight; 
			}else {
				parentNode.mLeft = fourthCase.mRight; 
			}
		}
		mNumPairs--; 
	}

	// makeEmpty()
	// re-sets D to the empty state.
	// pre: none
	public void makeEmpty() {
		deleteAll();
	}

	// toString()
	// pre: none
	// prints a text representation of D to the file pointed to by out
	public String toString() {
		return printInOrder(mRoot); 
	}
}